source/particle/generator/probability_density_function.cc

```diff
@@ -60,23 +60,30 @@ namespace aspect
         MPI_Scan(&local_function_integral, &start_weight, 1, MPI_DOUBLE, MPI_SUM, this->get_mpi_communicator());
         start_weight -= local_function_integral;
 
-        // Build the map between integrated probability density and cell
-        std::map<double, types::LevelInd> cell_weights;
-        typename DoFHandler<dim>::active_cell_iterator cell = this->get_dof_handler().begin_active(),
-                                                       endc = this->get_dof_handler().end();
-        for (unsigned int i = 0; cell!=endc; ++cell)
-          if (cell->is_locally_owned())
-            {
-              // adjust the cell_weights according to the local starting weight
-              cell_weights.insert(std::make_pair(accumulated_cell_weights[i], types::LevelInd(cell->level(),cell->index())));
-              ++i;
-            }
-
         // Calculate start id and number of local particles
         const types::particle_index start_id = llround(static_cast<double> (n_tracers)  * start_weight / global_function_integral);
         const types::particle_index n_local_particles = llround(static_cast<double> (n_tracers) * local_function_integral / global_function_integral);
 
-        generate_particles_in_subdomain(cell_weights,start_id,n_local_particles,particles);
+        // Uniform distribution on the interval [0,1]. This
+        // will be used to select cells for all particles.
+        boost::uniform_01<double> uniform_distribution_01;
+
+        // Loop over all particles to create locally and pick their cells
+        std::vector<unsigned int> particles_per_cell(this->get_triangulation().n_locally_owned_active_cells(),0);
+        for (types::particle_index current_particle_index = 0; current_particle_index < n_local_particles; ++current_particle_index)
+          {
+            // Draw the random number that determines the cell of the particle
+            const double random_weight =  accumulated_cell_weights.back() * uniform_distribution_01(this->random_number_generator);
```

Comment:
break this line.

Is there a way to just generate a random number in the range `[0,accumulated_cell_weights.back()]`? This would avoid the multiplication every time. (Not that it's going to make a runtime difference, the multiplication has to happen somewhere. I think it may just be simpler to read.)